crates/sui-core/src/safe_client.rs

```diff
@@ -359,6 +361,119 @@ where
         Ok(verified)
     }
 
+    fn verify_certificate_response_v3(
+        &self,
+        digest: &TransactionDigest,
+        HandleCertificateResponseV3 {
+            effects,
+            events,
+            input_objects,
+            output_objects,
+            auxiliary_data,
+        }: HandleCertificateResponseV3,
+    ) -> SuiResult<HandleCertificateResponseV3> {
+        let effects = self.check_signed_effects_plain(digest, effects, None)?;
+
+        // Check Events
+        match (&events, effects.events_digest()) {
+            (None, None) | (None, Some(_)) => {}
+            (Some(events), None) => {
+                if !events.data.is_empty() {
+                    return Err(SuiError::ByzantineAuthoritySuspicion {
+                        authority: self.address,
+                        reason: "Returned events but no event digest present in the signed effects"
+                            .to_string(),
+                    });
+                }
+            }
+            (Some(events), Some(events_digest)) => {
+                fp_ensure!(
+                    &events.digest() == events_digest,
+                    SuiError::ByzantineAuthoritySuspicion {
+                        authority: self.address,
+                        reason: "Returned events don't match events digest in the signed effects"
+                            .to_string()
+                    }
+                );
+            }
+        }
+
+        // Check Input Objects
+        if let Some(input_objects) = &input_objects {
+            let expected: HashMap<_, _> = effects
+                .old_object_metadata()
```

Comment:
We should use modified_at_versions to be consistent with how we fetch them in the authority side, even thought these two functions are almost identical.

Reply:
i needed to use `old_object_metadata` instead of `modified_at_versions` because this one includes the object digests (which need to be checked) vs just the object id and version